src/rules/default-import-match-filename.js

```diff
@@ -0,0 +1,120 @@
+import docsUrl from '../docsUrl'
+import isStaticRequire from '../core/staticRequire'
+
+/**
+ * @param {string} filename
+ * @returns {string}
+ */
+function removeExtension(filename) {
+  return filename.replace(/\.[^.]+$/, '')
+}
+
+/**
+ * @param {string} filename
+ * @returns {string}
+ */
+function normalizeFilename(filename) {
+  return filename.replace(/[-_.]/g, '').toLowerCase()
+}
+
+/**
+ * Test if local name matches filename.
+ * @param {string} localName
+ * @param {string} filename
+ * @returns {boolean}
+ */
+function isCompatible(localName, filename) {
+  const normalizedLocalName = localName.replace(/_/g, '').toLowerCase()
+
+  return (
+    normalizedLocalName === normalizeFilename(filename) ||
+    normalizedLocalName === normalizeFilename(removeExtension(filename))
+  )
+}
+
+/**
+ * Test if path starts with "./" or "../".
+ * @param {string} path
+ * @returns {boolean}
+ */
+function isLocalModule(path) {
+  return /^(\.\/|\.\.\/)/.test(path)
+}
+
+/**
+ * Get filename from a path.
+ * @param {string} path
+ * @returns {string | undefined}
+ */
+function getFilename(path) {
+  if (!isLocalModule(path)) return undefined
```

Comment:
why would we not want this enforced on things from node_modules? `import foo from 'package/bar'` should be an error.

Reply:
The gist is that this that rule can be annoying if you cannot control the filename, like when it a file in other peoples' package, which does not adhere to this rule.